test/functional/command_write_concern.test.js

```diff
@@ -17,245 +17,160 @@ var extend = function(template, fields) {
-          if (doc.ismaster) {
-            request.reply(firstSecondary[0]);
-          } else if (doc.endSessions) {
-            request.reply({ ok: 1 });
-          }
-        });
-
-        arbiterServer.setMessageHandler(request => {
-          var doc = request.document;
-          if (doc.ismaster) {
-            request.reply(arbiter[0]);
-          } else if (doc.endSessions) {
-            request.reply({ ok: 1 });
-          }
-        });
+      ]
+    };
+  }
+  decorateResponse(obj) {
+    Object.assign(this.responseDecoration, obj);
+  }
+  run(resultKey, testFn) {
+    const self = this;
+    co(function*() {
+      let primaryServer = yield mock.createServer(32000, 'localhost');
+      let firstSecondaryServer = yield mock.createServer(32001, 'localhost');
+      let arbiterServer = yield mock.createServer(32002, 'localhost');
+
+      primaryServer.setMessageHandler(request => {
+        var doc = request.document;
+        if (doc.ismaster) {
+          request.reply(self.serverStates.primary[0]);
+        } else if (doc[resultKey]) {
+          self.commandResult = doc;
+          request.reply(Object.assign({ ok: 1 }, self.responseDecoration));
+        } else if (doc.endSessions) {
+          request.reply({ ok: 1 });
+        }
+      });
 
-        var commandResult = null;
-        const client = configuration.newClient(
-          'mongodb://localhost:32000,localhost:32001,localhost:32002/test?replicaSet=rs'
-        );
+      firstSecondaryServer.setMessageHandler(request => {
+        var doc = request.document;
+        if (doc.ismaster) {
+          request.reply(self.serverStates.firstSecondary[0]);
+        } else if (doc.endSessions) {
+          request.reply({ ok: 1 });
+        }
+      });
 
-        client.connect(function(err, client) {
-          test.equal(null, err);
-          var db = client.db(configuration.db);
+      arbiterServer.setMessageHandler(request => {
+        var doc = request.document;
+        if (doc.ismaster) {
+          request.reply(self.serverStates.arbiter[0]);
+        } else if (doc.endSessions) {
+          request.reply({ ok: 1 });
+        }
+      });
 
-          db.collection('test')
-            .aggregate([{ $match: {} }, { $out: 'readConcernCollectionAggregate1Output' }], {
-              w: 2,
-              wtimeout: 1000
-            })
-            .toArray(function(err) {
-              test.equal(null, err);
-              test.deepEqual({ w: 2, wtimeout: 1000 }, commandResult.writeConcern);
+      const client = self.configuration.newClient(
+        'mongodb://localhost:32000,localhost:32001,localhost:32002/test?replicaSet=rs'
+      );
 
-              client.close(done);
-            });
-        });
+      client.connect(function(err, client) {
+        test.equal(null, err);
+        var db = client.db(self.configuration.db);
+        testFn(client, db);
       });
+    });
+  }
+}
+
+function withStubbedClient(testFn) {
```

Comment:
Are you going to further refactor this to use `sinon` or are we deferring that work to a later PR?

Reply:
Yeah I was planning to follow up by extracting `withStubbedClient` to `test/functional/shared.js` and making it use `sinon`.

Reply:
follow up in this PR or a subsequent one? Just trying to decide on whether to approve or not :)

Reply:
There's no rush to get this cleanup in so might as well do it in this PR, I'll mark it as a draft for now.

Reply:
Created a ticket to track this work: [NODE-2603](https://jira.mongodb.org/browse/NODE-2603)

Reply:
Given we're probably not getting to NODE-2603 particularly soon, I think it's a good idea to merge this PR in for now for the boilerplate reduction, so it doesn't sit around any longer. cc @nbbeeken